### ISSUE TITLE
Getting the exception thrown inside a closure

### ISSUE THREAD

**Issue by WyriHaximus** [krakjoe/parallel#16]
Been trying to get the exception thrown inside the closure passed into the runtime. But the resulting `parallel\Exception` contains no reference in `getPrevious` to it, was expecting it to be that way:

```php
try {
    $resolve($future->value());
} catch (\Throwable $throwable) {
    $reject($throwable->getPrevious());
}
```

**Comment by krakjoe** [comment]
The problem is that the exception is an object that was created in another thread - you can't access it from the parent. We could attempt to "save" some of the information from the exception in the child thread for retrieval in the parent, but saving the whole stack is pretty much out of the question (unreasonable).

I'll find some time to work on it, maybe today ...

**Comment by WyriHaximus** [comment]
Ah yes of course! Can work around that with [`wyrihaximus/json-throwable`](https://github.com/WyriHaximus/php-json-throwable) for the bigger projects. Having the message, line, file, code, and previous would be great. Having a (sanitized) trace would be even better but if this would be causing to much issues in this extension I'll have to do without.

**Comment by WyriHaximus** [comment]
Just a random idea but you could require exceptions to implement an interface to ensure safe transfer between threads.

**Comment by krakjoe** [comment]
To transfer an object from another thread is illegal in php, no matter the interface ... *share nothing* ...

It must be copied out of the context that created it, and into the context that wants to use it ... that's the only way ...

**Comment by WyriHaximus** [comment]
True and not trying to achieve that, but what I meant is that the interface could represent a safely serializable exception that can be put back together on the main thread side again by transferring as scalars.

**Comment by krakjoe** [comment]
Well an exception cannot know what is on the stack that generated it at the time of construction - and it's the trace that may have non-serializable variables on the stack .... To just copy everything but the trace we don't need any special interface, the problem arises when we try to copy the stack for the trace, and it contains non-serializable variables ...

**Comment by WyriHaximus** [comment]
What about stripping anything non-scalar from the trace? Just thinking out loud here :)

**Comment by krakjoe** [comment]
That's the kind of thing I'm thinking about doing, but not just stripping - if there's an array in the trace you want to know how many elements it had, if there's an object, you want to know it's class, and possibly it's handle number ... this is the sort of thing I'm thinking of doing, and then re-implementing getTraceAsString probably ...

I haven't had time to actually do anything today, it'll probably be tomorrow now ...

**Comment by WyriHaximus** [comment]
Awesome! Take your time, as always I rather wait longer and get something done well then get it soon and less polished :).

I'm also still working on automatic docker image building with this extension and several others in it.

**Comment by krakjoe** [comment]
I've started to work on this ... and then I had an idea ... still working on it ...

**Comment by WyriHaximus** [comment]
:tada: ! Looking forward to it, and let me know when there is something to test

**Comment by ghost** [comment]
FYI I've been exporting exceptions within PHP pthreads to the main thread by replacing any object with `Object(class_name)#object_id`. Arrays are per-se fine to serialize as long as its elements are non-objects.

https://github.com/CharlotteDunois/Phoebe/blob/550d46e83947b81c00180dd41dc802b0e5afe693/src/Message.php#L206-L216

**Comment by dktapps** [comment]
The only time I've ever had trouble serializing exceptions are when `Threaded` objects appeared in the stack trace, which isn't a problem in parallel.

**Comment by krakjoe** [comment]
@dktapps That's only by chance, and because of the specific nature of the code you are running, there are even core objects (like generators and closures) that cannot be serialized.

I don't intend to serialize anything in parallel, just to be clear ...

**Comment by dktapps** [comment]
that's good, serialize was an awful hack.

**Comment by WyriHaximus** [comment]
> that's good, serialize was an awful hack.

Hear hear!

**Comment by krakjoe** [comment]
Just to update this ... since we switched to using native frames, and implementing a lower level executor, we can now intercept exceptions without installing hooks or overriding instructions (which may not be forward compatible with jit) ...

I'm working on other (what I think are) more important things at the moment ... but this isn't forgotten about ...

**Comment by krakjoe** [comment]
I'm on this now ... but came across a bug in copying code, I need to deal with strings better ... wip ...

My current implementation turns Future into a container for uncaught exceptions and future values, it changes the rules a little - you can't ignore throw() from the task (must retain a future) ...

What I'm not sure of, is do we want methods on Future to access the exception (isError/getError), or do we want the exception to be thrown upon trying to access the value with value() method (only) ?

**Comment by dktapps** [comment]
Given the way Future::select() works, I think it would be cleaner to have a getError().

**Comment by WyriHaximus** [comment]
IMHO Making `Future::value()` throw the uncaught exception makes sense and feels most natural workflow wise. Because either you have your happy flow inside the try and your failureflow inside the catch

**Comment by krakjoe** [comment]
@dktapps there is no Future::select anymore :)

**Comment by dktapps** [comment]
Oops, I have some catching up to do apparently ...
I think it's probably not difficult to have both, right?

**Comment by WyriHaximus** [comment]
> Oops, I have some catching up to do apparently ...

The readme is the first thing I check after a new build in the morning to be sure I'm still up to date giving the high change rate on this so far awesome extension

**Comment by krakjoe** [comment]
@dktapps 

> I think it's probably not difficult to have both, right?

Events is far superior to select() ... just another small learning curve, but when you've learned how it works, you won't be wanting for select anymore ...

@WyriHaximus 

> giving the high change rate on this so far awesome extension

I'm sorry about the high rate of change, but I'm pretty determined to get it right this time, and I'd rather break stuff now, than be stuck with the results of bad decisions in the years to come ... 

It won't be long now before we have the API we want, an API we can call final ... I'm pretty sure of that ...

I'm very grateful for the constant stream of feedback from everyone ...

**Comment by dktapps** [comment]
@krakjoe I was referring to throw / getError(), it's not difficult to have both of those, right?

**Comment by WyriHaximus** [comment]
> @WyriHaximus
> 
> > giving the high change rate on this so far awesome extension
> 
> I'm sorry about the high rate of change, but I'm pretty determined to get it right this time, and I'd rather break stuff now, than be stuck with the results of bad decisions in the years to come ...
> 

IMHO don't be, you're doing an amazing job here and we're along for the ride

> It won't be long now before we have the API we want, an API we can call final ... I'm pretty sure of that ...
> 

Looking forward to that :tada: . Already have a small project in production with this and currently working on updating a second to use threads instead of child processes. Can't wait to see the results of that change

> I'm very grateful for the constant stream of feedback from everyone ...

Happy to help out :+1:

**Comment by krakjoe** [comment]
Implemented in 1485ffd8bda64d93756d8cadb50a078003c0c269

Where Future::value() is called uncaught exceptions shall be rethrown in the calling context, where an exception is of a class that the calling context does not have loaded a \parallel\Future\Error\Foreign exception will be thrown.

Where Events are used, the read event for a Future that would rethrow will have Event::$value instanceof Throwable.

Since there are only two ways to interact with the Future value, and since we can concisely describe how in both of those circumstances you detect an error, it doesn't seem necessary to add more methods to Future.
